internal/pkg/cli/storage_init.go

```diff
@@ -395,25 +395,105 @@ func (o *initStorageOpts) askStorageNameWithDefault(friendlyText, defaultName st
+	}
+	workload, err := o.configSel.Workload(storageInitSvcPrompt, "", o.appName)
+	if err != nil {
+		return fmt.Errorf("select a workload from app %s: %w", o.appName, err)
 	}
+	o.workloadName = workload
+	return nil
+}
+
+func (o *initStorageOpts) askLocalWorkload() error {
 	workload, err := o.sel.Workload(storageInitSvcPrompt, "")
 	if err != nil {
 		return fmt.Errorf("retrieve local workload names: %w", err)
 	}
 	o.workloadName = workload
+	o.workloadExists = true
+	return nil
+}
+
+func (o *initStorageOpts) validateOrAskLifecycle() error {
+	if o.lifecycle != "" {
+		return o.validateStorageLifecycle()
+	}
+	_, err := o.ws.ReadFile(o.ws.WorkloadAddonFileAbsPath(o.workloadName, fmt.Sprintf("%s.yml", o.storageName)))
+	if err == nil {
+		o.lifecycle = lifecycleWorkloadLevel
+		o.workloadExists = true
+		log.Infof("%s %s %s\n",
+			color.Emphasize("Lifecycle:"),
+			"workload-level",
+			color.Faint.Sprintf("(found %s)", o.ws.WorkloadAddonFilePath(o.workloadName, fmt.Sprintf("%s.yml", o.storageName))),
+		)
+		return nil
+	}
+	if _, ok := err.(*workspace.ErrFileNotExists); !ok {
+		return fmt.Errorf("check if %s addon exists for %s in workspace: %w", o.storageName, o.workloadName, err)
+	}
+	_, err = o.ws.ReadFile(o.ws.EnvAddonFileAbsPath(fmt.Sprintf("%s.yml", o.storageName)))
+	if err == nil {
+		o.lifecycle = lifecycleEnvironmentLevel
+		log.Infof("%s %s %s\n",
+			color.Emphasize("Lifecycle:"),
+			"environment-level",
+			color.Faint.Sprintf("(found %s)", o.ws.EnvAddonFilePath(fmt.Sprintf("%s.yml", o.storageName))),
+		)
+		return nil
+	}
+	if _, ok := err.(*workspace.ErrFileNotExists); !ok {
+		return fmt.Errorf("check if %s exists as an environment addon in workspace: %w", o.storageName, err)
+	}
+	options := []prompt.Option{
+		{
+			Value:        lifecycleWorkloadLevel,
+			FriendlyText: fmt.Sprintf(fmtLifecycleWorkloadFriendlyText, color.HighlightUserInput(o.workloadName)),
+		},
+		{
+			Value:        lifecycleEnvironmentLevel,
+			FriendlyText: lifecycleEnvironmentFriendlyText,
+		},
+	}
+	lifecycle, err := o.prompt.SelectOption(
+		fmt.Sprintf(fmtStorageInitLifecyclePrompt, o.workloadName),
+		"",
+		options,
+		prompt.WithFinalMessage("Lifecycle: "))
+	if err != nil {
+		return fmt.Errorf("ask for lifecycle: %w", err)
+	}
+	o.lifecycle = lifecycle
 	return nil
 }
 
-func (o *initStorageOpts) validateWorkloadName() error {
+func (o *initStorageOpts) validateStorageLifecycle() error {
+	for _, valid := range validLifecycleOptions {
+		if o.lifecycle == valid {
+			return nil
+		}
+	}
+	return fmt.Errorf("invalid lifecycle; must be one of %s", english.OxfordWordSeries(quoteStringSlice(validLifecycleOptions), "or"))
+}
+
+// validateWorkloadNameWithLifecycle requires the workload to be in the workspace if the storage lifecycle is on workload-level.
+// Otherwise, it only caches whether the workload is present.  
+func (o *initStorageOpts) validateWorkloadNameWithLifecycle() error {
+	if o.workloadExists {
```

Comment:
Can we possibly remove `o.workloadExists` and `o.cycle` and just use `o.workloadName`, `o.envName` to determine if the cycle is env level or workload level? If they pass workload name or env name through flags, in `Validate` or `Ask` we'll check if it exists. Otherwise, we ask if they want env level or workload level and if so which workload/env they want to use.

Reply:
I don't follow 🤔 . There is no `o.envName`.  Environment-level storage is not env-specific. Besides, it is entirely possible that people run
```
copilot storage int --workload api --lifecycle environment
```
in a workspace that does not have `api`.

Reply:
Ah I see.